include/ginkgo/core/config/registry.hpp

```diff
@@ -0,0 +1,238 @@
+struct base_type {};
+
+template <typename T>
+struct base_type<
+    T, typename std::enable_if<std::is_convertible<T*, LinOp*>::value>::type> {
+    using type = LinOp;
+};
+
+template <typename T>
+struct base_type<T, typename std::enable_if<
+                        std::is_convertible<T*, LinOpFactory*>::value>::type> {
+    using type = LinOpFactory;
+};
+
+template <typename T>
+struct base_type<T, typename std::enable_if<std::is_convertible<
+                        T*, stop::CriterionFactory*>::value>::type> {
+    using type = stop::CriterionFactory;
+};
+
+
+/**
+ * allowed_ptr is a type-erased object for LinOp/LinOpFactory/CriterionFactory
+ * shared_ptr.
+ */
+class allowed_ptr {
+public:
+    /**
+     * The constructor accept any shared pointer whose base type is LinOp,
+     * LinOpFactory, or CriterionFactory. We use template rather than
+     * constructor without template because it allows user to directly use
+     * uninitialized_list in registry constructor without wrapping allowed_ptr
+     * manually.
+     */
+    template <typename Type>
+    allowed_ptr(std::shared_ptr<Type> obj);
+
+    /**
+     * Check whether it contains the Type data
+     *
+     * @tparam Type  the checking type
+     */
+    template <typename Type>
+    bool contains() const;
+
+    /**
+     * Get the shared pointer with Type
+     *
+     * @tparam Type  the desired type
+     *
+     * @return the shared pointer of Type
+     */
+    template <typename Type>
+    std::shared_ptr<Type> get() const;
+
+private:
+    struct generic_container {
+        virtual ~generic_container() = default;
+    };
+
+    template <typename Type>
+    struct concrete_container : generic_container {
+        concrete_container(std::shared_ptr<Type> obj) : ptr{obj} {}
+
+        std::shared_ptr<Type> ptr;
+    };
+
+    std::shared_ptr<generic_container> data_;
+};
+
+
+template <typename Type>
+inline allowed_ptr::allowed_ptr(std::shared_ptr<Type> obj)
+{
+    data_ =
+        std::make_shared<concrete_container<typename base_type<Type>::type>>(
+            obj);
+}
+
+
+template <typename Type>
+inline bool allowed_ptr::contains() const
+{
+    return dynamic_cast<const concrete_container<Type>*>(data_.get());
+}
+
+
+template <typename Type>
+inline std::shared_ptr<Type> allowed_ptr::get() const
+{
+    GKO_THROW_IF_INVALID(this->template contains<Type>(),
+                         "does not hold the requested type.");
+    return dynamic_cast<concrete_container<Type>*>(data_.get())->ptr;
+}
+
+
+}  // namespace detail
+
+
+/**
+ * This class stores additional context for creating Ginkgo objects from
+ * configuration files.
+ *
+ * The context can contain user provided objects of the following types:
+ * - LinOp
+ * - LinOpFactory
+ * - CriterionFactory
+ *
+ * Additionally, users can provide mappings from a configuration (provided as
+ * a pnode) to user-defined types that are derived from one of the previous
+ * stated types.
+ */
+class registry {
```

Comment:
I'm questioning if this class needs to be defined in a public header. The only interaction from the user side should be to create an object of this type. Anything else, especially retrieving data from it, should not be done by the user. 
The only issue would be if users write their own mapping and want to retrieve something from the registry. But TBH, I'm not sure if that case would ever appear.

Perhaps it's best to explain my general reasoning for this (and also for the comment on the `search_data` function). If the user was able to put in and retrieve data then this object could be used solely for this task. Users could then just throw in any ginkgo object they might create, pass the registry around, and then just retrieve again the object. In that sense the registry acts as a kind of global state for ginkgo objects. If users really want to do that, they can do that easily by themselves, but we should not provide them any help in that.

I know that I'm arguing for restricting the allowed usage of this class by a lot. Generally, we should provide users with as much freedom as possible, but in this case I think it could enable bad-practices on the user side, so I would rather keep it as minimal as possible in this case.

Reply:
I have put the retrieve function into protected.
I would say the emplace is still neccessary, otherwise they need to keep the same stored data map with allowed_ptr somewhere to allow depenency build. I prefer they use registry than use allowed_ptr. When we do not allow the LinOp/CriterionFactory, reusability in the code, the emplace function is a way to handle it. Without it, it makes things annoying and make the factory not reducing the configuration work.

Reply:
The `registry_accessor` is a good solution. I was at first a bit concerned about putting the `search_data` into private, because I thought it would lead to many friend classes, but your approach is quite nice.
I'm fine with the emplace function staying, even though I can't really follow your argument here.